memory/src/main/scala/filodb.memory/data/OffheapLFSortedIDMap.scala

```diff
@@ -246,22 +294,151 @@ class OffheapLFSortedIDMapReader(memFactory: MemFactory, holderKlass: Class[_ <:
     innerBinSearch(0, mapLen)
   }
 
+  /**
+   * Acquire exclusive access to this map, spinning if necessary. Exclusive lock isn't re-entrant.
+   */
+  def acquireExclusive(inst: MapHolder): Unit = {
+    // Spin-lock implementation. Because the owner of the shared lock might be blocked by this
+    // thread as it waits for an exclusive lock, deadlock is possible. To mitigate this problem,
+    // timeout and retry, allowing shared lock waiters to make progress.
+
+    var timeoutNanos = 1000000 // 1 millisecond
+    var warned = false
+
+    while (true) {
+      if (tryAcquireExclusive(inst, timeoutNanos)) {
+        return
+      }
+
+      timeoutNanos = Math.min(timeoutNanos << 1, 1000000000) // limit timeout to 1 second
+
+      if (!warned && timeoutNanos >= 1000000000) {
+        _logger.warn(s"Waiting for exclusive lock: $inst")
+        warned = true
+      }
+    }
+  }
+
+  /**
+   * Acquire exclusive access to this map, spinning if necessary. Exclusive lock isn't re-entrant.
+   *
+   * @return false if timed out
+   */
+  private def tryAcquireExclusive(inst: MapHolder, timeoutNanos: Long): Boolean = {
+    // Spin-lock implementation.
+
+    var lockState = 0
+
+    // First set the high bit, to signal an exclusive lock request.
+
+    var done = false
+    do {
+      lockState = UnsafeUtils.getIntVolatile(inst, lockStateOffset)
+      if (lockState < 0) {
+        // Wait for exclusive lock to be released.
+        Thread.`yield`
+      } else if (UnsafeUtils.unsafe.compareAndSwapInt(inst, lockStateOffset, lockState, lockState | 0x80000000)) {
+        if (lockState == 0) {
+          return true
+        }
+        // Scala doesn't support a simple break statement for some reason.
+        done = true
+      }
+    } while (!done)
+
+    // Wait for shared lock owners to release the lock.
+
+    val endNanos = System.nanoTime + timeoutNanos
```

Comment:
Right now timeoutNanos defaults to 1ms, which seems way too low right?  Seems like we need to make it configurable

Reply:
It doubles each time, but the loop around it never gives up.